numpy/core/src/multiarray/array_coercion.c

```diff
@@ -541,10 +541,6 @@ PyArray_Pack(PyArray_Descr *descr, char *item, PyObject *value)
     }
     int res = cast_raw_scalar_item(tmp_descr, data, descr, item);
 
-    if (PyDataType_REFCHK(tmp_descr)) {
-        /* We could probably use move-references above */
-        PyArray_Item_XDECREF(data, tmp_descr);
-    }
```

Comment:
Well, I guess we could try to use `move_references=1`, but overall, I think this just needs to be `PyArray_ClearBuffer(tmp_descr, data, 0, 1, 1);`

Just removing it should leak memory in your String example.

(I am happy to not test this.  You are right, there is absolutely no way to test this without going through a lot of trouble to create another dtype with references at this time.)

Reply:
You're right about the memory leak, thanks for the suggestion. I think it only leaks memory in my case if the cast succeeds, which my test script didn't exercise.

Reply:
Ah, right. If the cast fails the temporary `data` is never filled (unless it fails oddly late during the actual cast).